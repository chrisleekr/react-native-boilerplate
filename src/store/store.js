import {configureStore} from '@reduxjs/toolkit';
import {combineReducers} from 'redux';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER
} from 'redux-persist';
import AsyncStorage from '@react-native-async-storage/async-storage';

// Axios
import setupInterceptors from '../service/setupInterceptors';

// Slices
import authReducer from './slice/authSlice';
import todoReducer from './slice/todoSlice';
import userReducer from './slice/userSlice';

const persistConfig = {
  key: 'root',
  version: 1,
  storage: AsyncStorage
};

const rootReducer = combineReducers({
  auth: authReducer,
  todo: todoReducer,
  user: userReducer
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = configureStore({
  reducer: persistedReducer,
  // middleware option needs to be provided for avoiding the error. ref: https://redux-toolkit.js.org/usage/usage-guide#use-with-redux-persist
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      // Since axios is throwing serializecheck error, so had to turn off.
      // serializableCheck: false
      // But let see how it goes with following list instead of turnning it off.
      serializableCheck: {
        ignoredActions: [
          FLUSH,
          REHYDRATE,
          PAUSE,
          PERSIST,
          PURGE,
          REGISTER,
          'auth/login/rejected',
          'auth/register/rejected',
          'auth/passwordResetRequest/rejected',
          'todo/list/rejected',
          'todo/addOne/rejected',
          'todo/updateOne/rejected',
          'todo/deleteOne/rejected',
          'user/me/rejected',
          'user/updateMe/rejected'
        ]
      }
    })
});

setupInterceptors(store);

export const persistor = persistStore(store);
export default store;
