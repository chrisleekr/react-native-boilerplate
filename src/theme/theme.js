import {Dimensions} from 'react-native';
import {colors} from './colors';
const {width, height} = Dimensions.get('window');

// Themes:- Can alter values here. Can only be consumed through Context (see useTheme.js file)
const themes = {
  light: {
    name: 'light',
    barStyle: 'light-content',
    color: '#000000',
    labelColor: '#525E75',
    invertColor: '#ffffff',
    layoutBg: '#f0f0f0',
    headerBg: '#f6f7f8',
    cardBg: '#e4dfa7',
    accent: '#0071ff',
    error: '#B00020',
    errorText: '#ed4337',
    lightBg: '#F5EE9E',
    borderColor: '#B00020',
    buttonActiveColor: '#6200EE',
    buttonDisabledColor: 'rgba(0, 0, 0, 0.26)',
    card: {
      shadowColor: '#000000',
      borderColor: '#B00020'
    },
    ...colors
  },
  dark: {
    name: 'dark',
    barStyle: 'dark-content',
    color: '#ffffff',
    labelColor: '#F1DDBF',
    invertColor: '#000000',
    layoutBg: '#121212',
    headerBg: '#1e1e1e',
    cardBg: '#3f4045',
    accent: '#0071ff',
    error: '#B00020',
    errorText: '#ed4337',
    lightBg: '#F5EE9E',
    borderColor: '#ffffff',
    buttonActiveColor: '#BB86FC',
    buttonDisabledColor: 'rgba(255,255,255, 0.26)',
    card: {
      shadowColor: '#121212',
      borderColor: '#ffffff'
    },
    ...colors
  }
};

const window = {
  // app dimensions
  width,
  height
};

export {themes, window};
