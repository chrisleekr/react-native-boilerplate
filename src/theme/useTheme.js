/**
 * For Dark/Light Mode
 * @exports ThemeContext, ThemeProvider, useTheme
 */

import React, {useState, createContext, useContext} from 'react';
import {Appearance} from 'react-native';

// Import preconfigured themes from theme file
import {themes} from './theme';

// Context
export const ThemeContext = createContext(null);

// Provider to be used in index/App/or top of any parent
export const ThemeProvider = ({children}) => {
  const colorScheme = Appearance.getColorScheme();

  const [theme, setTheme] = useState(themes[colorScheme]);
  return (
    <ThemeContext.Provider value={{theme, setTheme}}>
      {children}
    </ThemeContext.Provider>
  );
};

// useTheme hook for easy access to theme and setTheme
export const useTheme = () => {
  const {theme, setTheme} = useContext(ThemeContext);

  const toggleTheme = v => {
    setTheme(v ? themes.dark : themes.light);
  };

  return {theme, toggleTheme};
};
