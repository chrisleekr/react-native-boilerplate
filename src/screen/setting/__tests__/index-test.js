import React from 'react';
import * as redux from 'react-redux';
import {NavigationContainer} from '@react-navigation/native';
import {
  cleanup,
  fireEvent,
  render,
  waitFor,
  act
} from '@testing-library/react-native';

import {ThemeProvider} from '../../../theme/useTheme';

import store from '../../../store/store';

import SettingScreen from '../index';

let rendered;
let element;

const mockGoBack = jest.fn();

const component = (
  <ThemeProvider>
    <redux.Provider store={store}>
      <NavigationContainer>
        <SettingScreen navigation={{goBack: mockGoBack}} />
      </NavigationContainer>
    </redux.Provider>
  </ThemeProvider>
);

describe('SettingScreen', () => {
  afterEach(cleanup);

  beforeEach(async () => {
    rendered = render(component);
    await waitFor(() => expect(rendered.toJSON()).not.toBeNull());
  });

  it('matches snapshot', () => {
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  describe('goBack', () => {
    beforeEach(async () => {
      fireEvent.press(rendered.getByTestId('back-button'));
    });

    it('triggers mockGoBack', () => {
      expect(mockGoBack).toHaveBeenCalled();
    });
  });

  describe('dark mode', () => {
    beforeEach(() => {
      element = rendered.getByTestId('dark-mode-switch');
    });
    describe('check current state', () => {
      it('sets switch as unchecked', () => {
        element = rendered.getByTestId('dark-mode-switch');

        expect(element.props.value).toBeFalsy();
      });
    });
    describe('toggle dark mode', () => {
      beforeEach(async () => {
        act(() => {
          fireEvent(element, 'onValueChange', true);
        });
      });

      it('matches snapshot', () => {
        expect(rendered.toJSON()).toMatchSnapshot();
      });

      it('sets switch as checked', () => {
        expect(element.props.value).toBeTruthy();
      });
    });

    describe('toggle light mode', () => {
      beforeEach(async () => {
        act(() => {
          fireEvent(element, 'onValueChange', false);
        });
      });

      it('matches snapshot', () => {
        expect(rendered.toJSON()).toMatchSnapshot();
      });

      it('sets switch as unchecked', () => {
        expect(element.props.value).toBeFalsy();
      });
    });
  });
});
