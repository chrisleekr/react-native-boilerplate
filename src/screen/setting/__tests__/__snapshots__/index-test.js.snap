// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingScreen dark mode toggle dark mode matches snapshot 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "backgroundColor": "#121212",
      },
      Object {
        "flex": 1,
        "flexDirection": "column",
        "padding": 10,
      },
    ]
  }
  testID="setting-screen"
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "left": 4,
        "opacity": 1,
        "position": "absolute",
        "top": 10,
      }
    }
    testID="back-button"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontSize": 25,
          },
          undefined,
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      󱦱
    </Text>
  </View>
  <View
    style={
      Object {
        "marginTop": 25,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 21,
            "fontWeight": "bold",
            "marginBottom": 10,
          },
          Object {
            "color": "#ffffff",
          },
        ]
      }
    >
      Settings
    </Text>
    <RCTScrollView
      showsHorizontalScrollIndicator={true}
      showsVerticalScrollIndicator={false}
    >
      <View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 5,
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "padding": 10,
              },
              Object {
                "backgroundColor": "#3f4045",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 15,
                },
                Object {
                  "color": "#F1DDBF",
                },
              ]
            }
            testID="dark-mode-label"
          >
            Dark mode
          </Text>
          <RCTSwitch
            accessibilityRole="switch"
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            onTintColor="#81b0ff"
            style={
              Array [
                Object {
                  "height": 31,
                  "width": 51,
                },
                Object {
                  "backgroundColor": "#3e3e3e",
                  "borderRadius": 16,
                },
              ]
            }
            testID="dark-mode-switch"
            thumbTintColor="#f5dd4b"
            tintColor="#767577"
            value={true}
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
</RNCSafeAreaView>
`;

exports[`SettingScreen dark mode toggle light mode matches snapshot 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "backgroundColor": "#f0f0f0",
      },
      Object {
        "flex": 1,
        "flexDirection": "column",
        "padding": 10,
      },
    ]
  }
  testID="setting-screen"
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "left": 4,
        "opacity": 1,
        "position": "absolute",
        "top": 10,
      }
    }
    testID="back-button"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 25,
          },
          undefined,
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      󱦱
    </Text>
  </View>
  <View
    style={
      Object {
        "marginTop": 25,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 21,
            "fontWeight": "bold",
            "marginBottom": 10,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      Settings
    </Text>
    <RCTScrollView
      showsHorizontalScrollIndicator={true}
      showsVerticalScrollIndicator={false}
    >
      <View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 5,
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "padding": 10,
              },
              Object {
                "backgroundColor": "#e4dfa7",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 15,
                },
                Object {
                  "color": "#525E75",
                },
              ]
            }
            testID="dark-mode-label"
          >
            Dark mode
          </Text>
          <RCTSwitch
            accessibilityRole="switch"
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            onTintColor="#81b0ff"
            style={
              Array [
                Object {
                  "height": 31,
                  "width": 51,
                },
                Object {
                  "backgroundColor": "#3e3e3e",
                  "borderRadius": 16,
                },
              ]
            }
            testID="dark-mode-switch"
            thumbTintColor="#f4f3f4"
            tintColor="#767577"
            value={false}
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
</RNCSafeAreaView>
`;

exports[`SettingScreen matches snapshot 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "backgroundColor": "#f0f0f0",
      },
      Object {
        "flex": 1,
        "flexDirection": "column",
        "padding": 10,
      },
    ]
  }
  testID="setting-screen"
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "left": 4,
        "opacity": 1,
        "position": "absolute",
        "top": 10,
      }
    }
    testID="back-button"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 25,
          },
          undefined,
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      󱦱
    </Text>
  </View>
  <View
    style={
      Object {
        "marginTop": 25,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 21,
            "fontWeight": "bold",
            "marginBottom": 10,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      Settings
    </Text>
    <RCTScrollView
      showsHorizontalScrollIndicator={true}
      showsVerticalScrollIndicator={false}
    >
      <View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 5,
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "padding": 10,
              },
              Object {
                "backgroundColor": "#e4dfa7",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 15,
                },
                Object {
                  "color": "#525E75",
                },
              ]
            }
            testID="dark-mode-label"
          >
            Dark mode
          </Text>
          <RCTSwitch
            accessibilityRole="switch"
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            onTintColor="#81b0ff"
            style={
              Array [
                Object {
                  "height": 31,
                  "width": 51,
                },
                Object {
                  "backgroundColor": "#3e3e3e",
                  "borderRadius": 16,
                },
              ]
            }
            testID="dark-mode-switch"
            thumbTintColor="#f4f3f4"
            tintColor="#767577"
            value={false}
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
</RNCSafeAreaView>
`;
